Assignment 2.Rmd

```diff
@@ -95,6 +125,139 @@ Now practice with data from our class. Please create a **person-network** with t
+class<-class[which(class!="")]
+
+#select only 4 numeric elements in classcodde
+l<-unique(unlist(lapply(gsub('\\D+','', class),FUN=(function (x){substr(x,1,4)}))))
+
+mat<-unlist(apply(data_unique, 1,
+      function (x,classSeries){
+        unlist(lapply(classSeries,
+            function (i,x) {as.logical(grep(i,as.vector(x),perl=TRUE,value=FALSE)[1])}
+        ,x=x))
+      },classSeries=l))
+data_unique2<-data
+data_unique2[] <- lapply(data, as.character)
+
+# Uniform the classcode into one format
+for (j in 2:length(data_unique2[1,])) {
+  for (i in 1:length(data_unique2[,1])) {
+    rep<-which(unlist(lapply(l,grepl,x=data_unique2[i,j])))
+    if (length(rep)>0) {data_unique2[i,j]<-l[rep]}
+  }
+}
+
+# Person-class matrix
+data3 <- gather(data_unique2,classcode,value=class,2:7)
+
+data4<-data3[,-2]
+data5<- mutate(data4,count=1)
+attach(data5)
+data6 <- filter(data5, class != "" & class != "4050")
+data7 <- unique(data6)
+data8 <- spread(data7, key=class,value=count)
+data8[is.na(data8)]<-0
+row.names(data8)<-data8$abb
+data8<-data8[,-1]
+data9 <-as.matrix(data8)
+
+#person-person matrix
+data10<-data9 %*% t(data9)
+diag(data10)<-NA
+
+g4<-graph.adjacency(data10,mode="undirected",diag=FALSE)
+
+
+plot(g4, layout=layout.fruchterman.reingold)
+
+btns<-betweenness(g4)
+which(btns==max(btns))   
+dg <- degree(g4)
+which(dg==max(dg))      
+
+
+
+```
+
+
+Method2 (Use function to do the person-class mapping)
+```{r}
+library(tidyr)
+library(tidyverse)
+library(igraph)
+
+data <- read.csv("hudk4050-classes.csv")
+data$abb <- paste(data$First.Name, data$Last.Name)
+data <- data[c(9, 3,4,5,6,7,8)]
+attach(data)
+data_unique <- data[c(2:7)]
+class <- c(unique(as.vector(as.matrix(data_unique))))
+class<-class[which(class!="")]
+#select only 4 numeric elements in classcodde
+l<-unique(unlist(lapply(gsub('\\D+','', class),FUN=(function (x){substr(x,1,4)}))))
+l <- l[l!="4050"]
+
+#Convert classcode to uniform format
+mapClass <- function (x, class) {
+  item<-grep(paste(x,"+",sep=""), class, perl=TRUE, value=TRUE)[1]
+  ind<-regexpr(x,item)[1]
+  item<-substr(item,1,ind+nchar(x)-1)
+  if (ind==1) {return(item)}
+  if (substr(item,ind-1,ind-1) != " ") {
+    item<-paste(substr(item,1,ind-1),substr(item,ind,nchar(item)),sep=" ")
+  }
+  return(item)
+}
+cleanClass<-unlist(lapply(l,FUN=mapClass,class=class))
+
+
+# Person to class matrix
+mat<-unlist(apply(data_unique, 1,
+      function (x,classSeries){
+        unlist(lapply(classSeries,
+            function (i,x) {as.logical(grep(i,as.vector(x),perl=TRUE,value=FALSE)[1])}
+        ,x=x))
+      },classSeries=l))
+mat[is.na(mat)] = 0
+
+
+# Person to Person matrix
+p2p <- t(mat)%*%mat
+colnames(p2p)<-data$abb
+rownames(p2p)<-data$abb
+diag(p2p)<-NA
+
+g3<-graph.adjacency(p2p,mode="undirected",diag=FALSE)
+
+
+plot(g3, layout=layout.fruchterman.reingold)
+
+btns<-betweenness(g3)
+which(btns==max(btns))   
+dg <- degree(g3)
+which(dg==max(dg))      
```

Comment:
Great job overall! You demonstrated excellent coding skills. Try to interpret the centrality measures within the context of this dataset.